sprint_8-solution-time-and-duration

A-Glebov opened the pull request:
1. Добавили новые поля Duration duration, LocalDateTime starTime, метод для расчета конечного времени задачи

2. В Epic добавили поле endTime для расчета конечного времени
3. Доработали опцию сохранения состояния в файл с учетом новых полей
4. Добавили проверку на пересечение задач во времени
5. Добавили метод getPrioritizedTask для сортировки задач по приоритету
6. Написали тесты для нового функционала программы

A-Glebov committed:
1. Добавили новые поля Duration duration, LocalDateTime starTime, метод для расчета конечного времени задачи
2. В Epic добавили поле endTime для расчета конечного времени
3. Доработали опцию сохранения состояния в файл с учетом новых полей
4. Добавили проверку на пересечение задач во времени
5. Добавили метод getPrioritizedTask для сортировки задач по приоритету
6. Написали тесты для нового функционала программы

A-Glebov committed:
1. Добавили новые поля Duration duration, LocalDateTime starTime, метод для расчета конечного времени задачи
2. В Epic добавили поле endTime для расчета конечного времени
3. Доработали опцию сохранения состояния в файл с учетом новых полей
4. Добавили проверку на пересечение задач во времени
5. Добавили метод getPrioritizedTask для сортировки задач по приоритету
6. Написали тесты для нового функционала программы

A-Glebov committed:
1. Добавили новые поля Duration duration, LocalDateTime starTime, метод для расчета конечного времени задачи
2. В Epic добавили поле endTime для расчета конечного времени
3. Доработали опцию сохранения состояния в файл с учетом новых полей
4. Добавили проверку на пересечение задач во времени
5. Добавили метод getPrioritizedTask для сортировки задач по приоритету
6. Написали тесты для нового функционала программы

BerezinYaroslav commented on src/service/TaskManager.java:
```diff
@@ -74,8 +75,13 @@ public interface TaskManager {
     // Получение истории просмотров
     List<Task> getHistory();
+
+    TreeSet<Task> getPrioritizedTasks();
```
⛔️ Этот метод должен возвращать именно список задач

BerezinYaroslav replied:
✅

BerezinYaroslav commented on src/service/InMemoryTaskManager.java:
```diff
@@ -16,6 +19,64 @@ public class InMemoryTaskManager implements TaskManager {
+    }
+
+
+    public boolean checkTaskForIntersectionInTime(Task task, Task existTask) {
```
✅ Верно

BerezinYaroslav commented on src/service/InMemoryTaskManager.java:
```diff
@@ -16,6 +19,64 @@ public class InMemoryTaskManager implements TaskManager {
+    }
+
+    @Override
+    public boolean validateTaskByDateTime(Task task) {
```
⛔️ Код-стайл стримов тоже следует поддерживать: 

```
return getPrioritizedTasks().stream()
        .anyMatch(existTask -> checkTaskForIntersectionInTime(task, existTask));
```

BerezinYaroslav replied:
✅

BerezinYaroslav commented on src/service/InMemoryTaskManager.java:
```diff
@@ -16,6 +19,64 @@ public class InMemoryTaskManager implements TaskManager {
+    @Override
+    public TreeSet<Task> getPrioritizedTasks() {
+
+        TreeSet<Task> prioritizedTasks = new TreeSet<>(Comparator
```
⛔️ Сложность выполнения этого метода должна быть O(n). Сортировка должна выполняться при добавлении новой задачи

BerezinYaroslav replied:
✅

BerezinYaroslav commented on test/model/EpicTest.java:
```diff
@@ -1,38 +1,134 @@
-import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.*;
 
 public class EpicTest {
```
✅ Для эпика тестов достаточно

BerezinYaroslav commented on test/service/InMemoryTaskManagerTest.java:
```diff
@@ -12,23 +14,37 @@
 public class InMemoryTaskManagerTest {
```
⛔️ Чтобы избежать дублирования кода, необходим базовый класс с тестами на каждый метод в интерфейсе TaskManagerTest<? extends TaskManager>. 

Для проверки каждого типа задач можно вынести общий функционал в базовый класс BaseTaskManager<? extends Task>, а все реализации наследовать от него.

Для каждого метода нужно проверить его работу:

a. Со стандартным поведением.
b. С пустым списком задач.
c. С неверным идентификатором задачи (пустой и/или несуществующий идентификатор).

BerezinYaroslav replied:
✅

BerezinYaroslav commented on test/service/InMemoryHistoryManagerTest.java:
```diff
@@ -22,9 +24,14 @@ public class InMemoryHistoryManagerTest {
     public void init() {
```
✅ Для HistoryManager тоже тестов вполне достаточно

BerezinYaroslav commented on test/service/FileBackedTaskManagerTest.java:
```diff
@@ -1,20 +1,21 @@
 
 import static org.junit.jupiter.api.Assertions.*;
 
 public class FileBackedTaskManagerTest {
```
✅ Для этого класса тоже тестов достаточно

BerezinYaroslav reviewed (changes requested):
Алексей, привет!

Хорошая работа, учитывая ее объем и сложность. Есть несколько важных моментов, которые нужно поправить, в остальном все отлично. 

Буду ждать твою новую работу 🤝

A-Glebov committed:
1. Добавили базовый абстрактный класс TaskManagerTest вынесли в него тестовые методы общие для всех менеджеров
2. Изменили тип возвращаемого значения метода List<Task> getPrioritizedTasks(Task task)
3. Добавили поле TreeSet<Task> prioritizedTasks

A-Glebov committed:
1. Добавили базовый абстрактный класс TaskManagerTest вынесли в него тестовые методы общие для всех менеджеров
2. Изменили тип возвращаемого значения метода List<Task> getPrioritizedTasks(Task task)
3. Добавили поле TreeSet<Task> prioritizedTasks

A-Glebov committed:
1. Добавили базовый абстрактный класс TaskManagerTest вынесли в него тестовые методы общие для всех менеджеров
2. Изменили тип возвращаемого значения метода List<Task> getPrioritizedTasks(Task task)
3. Добавили поле TreeSet<Task> prioritizedTasks

A-Glebov committed:
1. Добавили базовый абстрактный класс TaskManagerTest вынесли в него тестовые методы общие для всех менеджеров
2. Изменили тип возвращаемого значения метода List<Task> getPrioritizedTasks(Task task)
3. Добавили поле TreeSet<Task> prioritizedTasks

BerezinYaroslav commented on src/service/InMemoryTaskManager.java:
```diff
@@ -5,17 +5,77 @@
+
+    @Override
+    public List<Task> getPrioritizedTasks(Task task) {
+        prioritizedTasks.add(task);
```
⛔️ Все верно, но добавление задачи в список приоритета никак не связано с вызовом метода `getPrioritizedTasks`. Вызывать `prioritizedTasks.add(task);` ты должен в методе создания тасков, а этот метод вызывать не надо, только при получении списка задач по приоритету

BerezinYaroslav replied:
✅

BerezinYaroslav reviewed (changes requested):
Алексей, привет! 

Почти все поправил, нужно решить еще один момент и работа будет принята. Если что - пиши)

Удачи 🤝

A-Glebov committed:
1. метод getPrioritized теперь возвращает список задач по приоритету.
2. Сортировка осуществляется при добавлении задачи в менеджер

BerezinYaroslav reviewed (approved):
Алексей, привет! Принято!
